crucible-llvm/src/Lang/Crucible/LLVM/SimpleLoopInvariant.hs

```diff
@@ -0,0 +1,1151 @@
+                    }
+          return $ C.RegEntry (C.LLVMPointerRepr w) $ C.LLVMPointer havoc_blk (C.llvmPointerOffset (C.regValue left))
+
+        Nothing -> do
+          havoc_var <- liftIO $ W4.freshConstant sym (W4.safeSymbol "havoc_var") (W4.BaseBVRepr w)
+          put subst{ varHavoc = MapF.insert havoc_var (Const ()) (varHavoc subst) }
+          return $ C.RegEntry (C.LLVMPointerRepr w) $ C.LLVMPointer havoc_blk havoc_var
+
+  C.BoolRepr
+    | List.isPrefixOf "cmacaw" (show $ W4.printSymExpr $ C.regValue left) -> do
+      liftIO $ ?logMessage "SimpleLoopInvariant.joinRegEntry: cmacaw_reg"
+      return left
+    | otherwise -> do
+      -- liftIO $ ?logMessage $
+      --   "SimpleLoopInvariant.joinRegEntry: BoolRepr:"
+      --   ++ show (W4.printSymExpr $ C.regValue left)
+      --   ++ " \\/ "
+      --   ++ show (W4.printSymExpr $ C.regValue right)
+      join_varaible <- liftIO $ W4.freshConstant sym (W4.safeSymbol "macaw_reg") W4.BaseBoolRepr
+      return $ C.RegEntry C.BoolRepr join_varaible
+
+  C.StructRepr field_types -> do
+    liftIO $ ?logMessage "SimpleLoopInvariant.joinRegEntry: StructRepr"
+    C.RegEntry (C.regType left) <$> fmapFC (C.RV . C.regValue) <$> joinRegEntries sym
+      (Ctx.generate (Ctx.size field_types) $ \i ->
+        C.RegEntry (field_types Ctx.! i) $ C.unRV $ (C.regValue left) Ctx.! i)
+      (Ctx.generate (Ctx.size field_types) $ \i ->
+        C.RegEntry (field_types Ctx.! i) $ C.unRV $ (C.regValue right) Ctx.! i)
+  _ -> fail $ "SimpleLoopInvariant.joinRegEntry: unsupported type: " ++ show (C.regType left)
+
+
+applySubstitutionRegEntries ::
+  C.IsSymInterface sym =>
+  sym ->
+  MapF (W4.SymExpr sym) (W4.SymExpr sym) ->
+  Ctx.Assignment (C.RegEntry sym) ctx ->
+  Ctx.Assignment (C.RegEntry sym) ctx
+applySubstitutionRegEntries sym substitution = fmapFC (applySubstitutionRegEntry sym substitution)
+
+applySubstitutionRegEntry ::
+  C.IsSymInterface sym =>
+  sym ->
+  (MapF (W4.SymExpr sym) (W4.SymExpr sym)) ->
+  C.RegEntry sym tp ->
+  C.RegEntry sym tp
+applySubstitutionRegEntry sym substitution entry = case C.regType entry of
+  C.LLVMPointerRepr{} ->
+    entry
+      { C.regValue = C.LLVMPointer
+          (C.llvmPointerBlock (C.regValue entry))
+          (MapF.findWithDefault
+            (C.llvmPointerOffset (C.regValue entry))
+            (C.llvmPointerOffset (C.regValue entry))
+            substitution)
+      }
+  C.BoolRepr ->
+    entry
+  C.StructRepr field_types ->
+    entry
+      { C.regValue = fmapFC (C.RV . C.regValue) $
+          applySubstitutionRegEntries sym substitution $
+          Ctx.generate (Ctx.size field_types) $
+          \i -> C.RegEntry (field_types Ctx.! i) $ C.unRV $ (C.regValue entry) Ctx.! i
+      }
+  _ -> C.panic "SimpleLoopInvariant.applySubstitutionRegEntry" ["unsupported type: " ++ show (C.regType entry)]
+
+
+loadMemJoinVariables ::
+  (C.IsSymBackend sym bak, C.HasPtrWidth 64, C.HasLLVMAnn sym, ?memOpts :: C.MemOptions) =>
```

Comment:
Judging from the use of `HasPtrWidth 64`, does this only work for 64-bit? If so, it would be worth documenting this requirement somewhere.

Reply:
Yes, only on 64-bit. Agreed, should be documented.